### PR TITLE
contracts: fix CBOR unsigned int decoding in Subcall.sol

### LINKED ISSUE #323: `_parseCBORUint` failing in `_decodeReceiptUndelegateStart`
The `consensusTakeReceiptUndelegateStart` function in `Subcall.sol` is failing on testnet with the `InvalidLength` error.

### PR DESCRIPTION
**Pull request by CedarMist**
fixes #323

### CODE REVIEW

**Comment by CedarMist** [comment]
Ok, minor last thing.

I need to make the subcall stuff not error out if it encounters and unknown key.

File: contracts/contracts/Subcall.sol [hunk 1]
```diff
@@ -182,11 +188,37 @@ library Subcall {
...
     {
-        if (result[offset] & 0x40 != 0x40) revert InvalidLength();
+        uint8 prefix = uint8(result[offset]);
+        uint256 len;
 
-        uint256 len = uint8(result[offset++]) ^ 0x40;
+        if (prefix <= 0x17) {
+            return (offset + 1, prefix);
+        }
+        // byte array, parsed as a big-endian integer
```
Comment (matevz) [inline contracts/contracts/Subcall.sol:197]:
```suggestion
        // Byte array, parsed as a big-endian integer.
```

File: contracts/contracts/Subcall.sol [hunk 2]
```diff
@@ -182,11 +188,37 @@ library Subcall {
...
+        else if (prefix & 0x40 == 0x40) {
+            len = uint8(result[offset++]) ^ 0x40;
+        }
+        // unsigned integer, CBOR encoded
```
Comment (matevz) [inline contracts/contracts/Subcall.sol:201]:
```suggestion
        // Unsigned integer, CBOR encoded.
```

File: contracts/contracts/Subcall.sol [hunk 3]
```diff
@@ -63,6 +63,12 @@ library Subcall {
     /// CBOR parser expected a key, but it was not found in the map!
     error MissingKey();
 
+    /// Valid cannot be parsed as a uint
```
Comment (matevz) [inline contracts/contracts/Subcall.sol:66]: ?

**Review by matevz — APPROVED** [review]

**Review by matevz — COMMENTED** [review]
FYI this section https://www.rfc-editor.org/rfc/rfc8949.html#name-tag-validity mentions that the 128- and 256-bit integers are encoded as 0x40 (byte arrays).

**Comment by matevz** [comment]
> FYI this section https://www.rfc-editor.org/rfc/rfc8949.html#name-tag-validity mentions that the 128- and 256-bit integers are encoded as 0x40 (byte arrays).

Having the _cborParseUInt() split into 64-bit and smaller uints and 128-bit and greater uints codepaths opens potential gas attacks? Should we use gas padding to mitigate this?

**Comment by CedarMist** [comment]
> > FYI this section https://www.rfc-editor.org/rfc/rfc8949.html#name-tag-validity mentions that the 128- and 256-bit integers are encoded as 0x40 (byte arrays).
> 
> Having the _cborParseUInt() split into 64-bit and smaller uints and 128-bit and greater uints codepaths opens potential gas attacks? Should we use gas padding to mitigate this?

I'd prefer if the caller of the receipt parser does the gas padding if it's deemed necessary rather than building it into the parser itself. But yes, I'll add a note in the code about it.

File: contracts/contracts/Subcall.sol [hunk 4]
```diff
@@ -182,11 +188,39 @@ library Subcall {
...
+                len = 1;
+            } else if (prefix == 0x19) {
+                len = 2;
+            } else if (prefix == 0x1a) {
+                len = 4;
+            } else if (prefix == 0x1b) {
+                len = 8;
+            } else if (prefix == 0x1d) {
+                len = 16;
+            } else {
```
Comment (lubej) [inline contracts/contracts/Subcall.sol:213]: I see now, that this is actually wrong https://www.rfc-editor.org/rfc/rfc8949.html#name-jump-table-for-initial-byte

File: contracts/contracts/Subcall.sol [hunk 5]
```diff
@@ -170,6 +176,11 @@ library Subcall {
             (uint64, bytes)
         );
 
+        // 0xf6 = null, returns null in case receiptId not found
```
Comment (lubej) [inline contracts/contracts/Subcall.sol:179]: Also returns null(`0xf6`) in case epoch in not reached yet.

**Pull request merged**